src/internal/Observable.ts

```diff
@@ -392,4 +392,123 @@ export class Observable<T> implements Subscribable<T> {
+   *
+   * ## Example
+   *
+   * ```ts
+   * import { interval } from 'rxjs';
+   *
+   * async function main() {
+   *  // Subscribe to the observable using for await.
+   *  for await (const value of interval(1000)) {
+   *    console.log(value);
+   *
+   *    if (value > 5) {
+   *      // Unsubscribe from the interval if we get a value greater than 5
+   *      break;
+   *    }
+   *  }
+   * }
+   *
+   * main();
+   * ```
+   */
+  [Symbol.asyncIterator](): AsyncGenerator<T, void, void> {
+    let subscription: Subscription | undefined;
+    let hasError = false;
+    let error: any;
+    let completed = false;
+    const values: T[] = [];
+    const deferreds: [(value: any) => void, (reason: any) => void][] = [];
+
+    const handleError = (err: any) => {
+      hasError = true;
+      error = err;
+      while (deferreds.length) {
+        deferreds.shift()![1](err);
+      }
+    };
+
+    const handleComplete = () => {
+      completed = true;
+      while (deferreds.length) {
+        deferreds.shift()![0]({ value: undefined, done: true });
+      }
+    };
+
+    return {
+      next: (): Promise<IteratorResult<T>> => {
+        if (!subscription) {
+          // We only want to start the subscription when the user starts iterating.
+          subscription = this.subscribe({
+            next: (value) => {
+              if (deferreds.length) {
+                deferreds.shift()![0]({ value, done: false });
+              } else {
+                values.push(value);
+              }
+            },
+            error: handleError,
+            complete: handleComplete,
+          });
+        }
+
+        // If we already have some values in our buffer, we'll return the next one.
+        if (values.length) {
+          return Promise.resolve({ value: values.shift()!, done: false });
+        }
+
+        // This was already completed, so we're just going to return a done result.
+        if (completed) {
+          return Promise.resolve({ value: undefined, done: true });
+        }
+
+        // There was an error, so we're going to return an error result.
+        if (hasError) {
+          return Promise.reject(error);
+        }
+
+        // Otherwise, we need to make them wait for a value.
+        return new Promise((resolve, reject) => {
+          deferreds.push([resolve, reject]);
+        });
+      },
+      throw: (err): Promise<IteratorResult<T>> => {
+        subscription?.unsubscribe();
+        // NOTE: I did some research on this, and as of Feb 2023, Chrome doesn't seem to do
+        // anything with pending promises returned from `next()` when `throw()` is called.
+        // However, for consumption of observables, I don't want RxJS taking the heat for that
+        // quirk/leak of the type. So we're going to reject all pending promises we've nexted out here.
+        handleError(err);
+        return Promise.resolve({ value: undefined, done: true });
```

Comment:
it should reject on throw?

```suggestion
        return Promise.reject(err);
```

Reply:
No. The promise you get back from an async generator `throw` call doesn't normally reject.

Reply:
I might be misunderstanding, but I think an async generator should reject according to [AsyncGenerator.prototype.throw](https://tc39.es/ecma262/multipage/control-abstraction-objects.html#sec-asyncgenerator-prototype-throw)? 🧐

Also if `throw` does not reject, the catch clause in this test case will be skipped! (I'm not sure if this is intended though)
https://github.com/benlesh/rxjs/blob/e7968fde17cffc3ff35cf770c3d7f794af1053a5/spec/Observable-spec.ts#L1425-L1445

Forgot to say thank you for amazing work!! This feature is what I've been waiting for so long... 💜 RxJS

Reply:
Revisiting this and testing in Chrome console:

```ts
async function* myGen() {
   while (true) {
      yield 'hi'
   }
}

const g = myGen()

g.throw(new Error('wee')).catch(err => {
  console.log(`The throw call returned a rejected promise: ${err}`)
})
```

The throw call does indeed return a rejected error in this case.

Thanks for catching this. Sorry for the confusion.